crypto/rand/rand_uniform.c

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "crypto/rand.h"
+#include "internal/common.h"
+
+/*
+ * Implementation an optimal random integer in a range function.
+ * Refer: https://github.com/apple/swift/pull/39143 for a description
+ * of the algorithm.
+ */
+uint32_t ossl_rand_uniform_uint32(OSSL_LIB_CTX *ctx, uint32_t upper, int *err)
+{
+    uint32_t i, f;      /* integer and fractional parts */
+    uint32_t f2, rand;  /* extra fractional part and random material */
+    uint64_t prod;      /* temporary holding double width product */
+    const int max_followup_iterations = 10;
+    int j;
+
+    if (!ossl_assert(upper > 0)) {
+        *err = 0;
+        return 0;
+    }
+    if (unlikely(upper == 1))
+        return 0;
+    /* Get 32 bits of entropy */
+    if (RAND_bytes_ex(ctx, (unsigned char *)&rand, sizeof(rand), 0) <= 0) {
+        *err = 1;
+        return 0;
+    }
+    prod = (uint64_t)upper * rand;
+    i = prod >> 32;
+    f = prod & 0xffffffff;
+    if (likely(f <= 1 + ~upper))    /* 1+~upper == -upper but compilers whine */
+        return i;
+
+    for (j = 0; j < max_followup_iterations; j++) {
+        if (RAND_bytes_ex(ctx, (unsigned char *)&rand, sizeof(rand), 0) <= 0) {
```

Comment:
Just a wild thought, in view of the timing comments https://github.com/apple/swift/pull/39143#issuecomment-912185936: 

Since the block size of our DRBG is 128 bits, one could make a single generate call for 128 bits, which would yield input for an unconditional calculation using a 32bit integer and 96bit fractional part (up to 3 followup iterations in your wording). Wouldn't that be good enough in practice to avoid further looping?

Reply:
If we could guarantee a 128 bit integer type we could do this.
We need an integer that can be multiplied by the range and stay within limits.
IMO: 32 bits is the best we can hope for.

Although we require 64 bit integers we do not mandate 128 bit multiplication results.  Most modern compilers support such and cannot rely on it being available.  IMO, the time critical part of this code will be the random generator not the reduction to range.

Reply:
I wasn't thinking about using larger integers, only about joining the first four `RAND_bytes()` calls into a single call . But since the 128bits are cached internally and unused bits are consumed later iirc, this optimization might not bring such a large improvement as I initially thought.

Reply:
I misunderstood.

The DRBG cache their output, so I think you're correct in that that change wouldn't help much.

We don't actually want lots of bits.  The probability of needing a third block of 32 bits is under 2¯³³.  The probability of needing the second block is at most ½ and usually far less.